host: gatt: Support deferred Read and Write Response handling.

zhongzhijie1 opened the pull request:
Now stack handles read_request with a sync callback. It needs value and length to be returned right away.

This patch enhances the GATT server implementation to support deferred response handling for both Read and Write operations. It enables applications to perform long-running or context-dependent attribute operations without blocking the protocol layer.

Signed-off-by: Zhijie Zhong <zhongzhijie1@xiaomi.com>

zhongzhijie1 committed:
host: gatt: Support deferred Read and Write Response handling

Now the stack handles read_request with a synchronous callback which
requires the value and length to be returned immediately.

This patch enhances the GATT server implementation to support deferred
response handling for both Read and Write operations. It enables
applications to perform long-running or context-dependent attribute
operations without blocking the protocol layer.

Signed-off-by: Zhijie Zhong <zhongzhijie1@xiaomi.com>

alwa-nordic reviewed (changes requested):
- I'm missing an update to the APIs `bt_gatt_attr_read_func_t` and `bt_gatt_attr_write_func_t`.

 - I'm missing an explanation for why `bt_gatt_send_read_response` and `bt_gatt_send_write_response` take a `handle` argument.

The above are the reasons for my NAK. The rest can be discussed. The discussion would benefit from a clearly updated API documentation. Points to address:

 - This API change is breaking functionality. Currently `bt_gatt_attr.read` may be invoked locally.

 - Are we going to be breaking backwards compatibility?

 - The change is not compatible with our implementation when EATT is enabled.

alwa-nordic reviewed (changes requested):
I think this is a nice prototype PR. Let's plan how to proceed.

I have added some discussion points on the API I would like us to address.

I think at this point, we should add automated tests to check the implementation against the API.

zhongzhijie1 commented on include/zephyr/bluetooth/gatt.h:
```diff
@@ -156,6 +156,11 @@ struct bt_gatt_attr;
+ *  is deferred and the application must later send the response
+ *  using @ref bt_gatt_send_read_response(). Deferred responses
+ *  are only supported when Enhanced ATT (EATT) is not active.
+ *
```
Let's see:

Line 5168–5173

```
#if defined(CONFIG_BT_EATT)
	if (bt_eatt_count(conn) > 0) {
		LOG_WRN("EATT active: async response not supported");
		return -ENOTSUP;
	}
#endif
```


and

Line 1405–1410

```
		if (read == -EINPROGRESS) {
#if defined(CONFIG_BT_EATT)
			if (bt_att_is_enhanced(chan)) {
				LOG_WRN("Async read not supported on EATT channel");
				return BT_ATT_ERR_NOT_SUPPORTED;
			}
#endif
```


For user side, if bt_gatt_attr_read_func_t() returns -EINPROGRESS on EATT, it will fail.
send_read_response also checks this.
So I think user does not need to check IS_ENABLED(CONFIG_BT_EATT) in their callback.

zhongzhijie1 commented on include/zephyr/bluetooth/gatt.h:
```diff
@@ -199,6 +204,11 @@ typedef ssize_t (*bt_gatt_attr_read_func_t)(struct bt_conn *conn,
+ *  is deferred and the application must later send the response
+ *  using @ref bt_gatt_send_write_response(). Deferred responses
+ *  are only supported when Enhanced ATT (EATT) is not active.
+ *
```
```
 *  @note The application must send the response before the ATT timeout,
 *  otherwise the bearer will be terminated.
```

PavelVPV commented on include/zephyr/bluetooth/gatt.h:
```diff
@@ -156,6 +156,11 @@ struct bt_gatt_attr;
+ *  is deferred and the application must later send the response
+ *  using @ref bt_gatt_send_read_response(). Deferred responses
+ *  are only supported when Enhanced ATT (EATT) is not active.
+ *
```
>But for EATT, shall we allow user to store the chan pointer,
so they can send response to the specific EATT channel later?

Perhaps index, not pointer?

> Maybe we can wrap the callback parameters into a struct,
so it’s easier to extend for EATT or local read.

I'd try this.

zhongzhijie1 replied:
like `request_id`

PavelVPV commented on subsys/bluetooth/host/att.c:
```diff
@@ -2121,6 +2152,22 @@ static uint8_t att_write_rsp(struct bt_att_chan *chan, uint8_t req, uint8_t rsp,
+		}
+#endif
+		net_buf_unref(data.buf);
+		/* Async write: release buffer, app send response later */
```
> From the spec, for normal ATT, a new request shall not be sent before previous response.
So this case cannot happen on single bearer.

Yeah, but I mean there can be malicious peer. Host needs to be protected here.

> But for EATT (multiple bearers), it can happen, so we may need handle it.

still no more than 1 request per bearer.

Thalley commented:
I guess this PR is an attempt to solve this old issue? https://github.com/zephyrproject-rtos/zephyr/issues/29270

Thalley commented on include/zephyr/bluetooth/gatt.h:
```diff
@@ -199,6 +204,11 @@ typedef ssize_t (*bt_gatt_attr_read_func_t)(struct bt_conn *conn,
+ *  @note If this function returns ``-EINPROGRESS``, the write operation
+ *  is deferred and the application must later send the response
+ *  using @ref bt_gatt_send_write_response(). Deferred responses
+ *  are only supported when Enhanced ATT (EATT) is not active.
```
`when Enhanced ATT (EATT) is not active.` - What does this mean specifically?

`CONFIG_BT_EATT=n`? Or just that no EATT bearers are currently connected for the specific conn?

Thalley commented on subsys/bluetooth/host/gatt.c:
```diff
@@ -5155,6 +5155,64 @@ int bt_gatt_write_without_response_cb(struct bt_conn *conn, uint16_t handle,
+#endif
+
+	buf = bt_att_create_pdu(conn, BT_ATT_OP_READ_RSP, length);
+	if (!buf) {
```
```suggestion
	if (buf == NULL) {
```

Thalley commented on subsys/bluetooth/host/gatt.c:
```diff
@@ -5155,6 +5155,64 @@ int bt_gatt_write_without_response_cb(struct bt_conn *conn, uint16_t handle,
+#endif
+
+	buf = bt_att_create_pdu(conn, BT_ATT_OP_WRITE_RSP, 0);
+	if (!buf) {
```
```suggestion
	if (buf == NULL) {
```

Thalley commented on include/zephyr/bluetooth/gatt.h:
```diff
@@ -2073,6 +2102,24 @@ struct bt_gatt_write_params {
+ * @retval -ENOMEM Out of memory.
+ * @retval -ENOTSUP EATT active and deferred responses not supported.
+ */
+int bt_gatt_send_write_response(struct bt_conn *conn);
```
Or should we have a `bt_gatt_send_error_response`?

zhongzhijie1 committed:
host: gatt: Use attribute handle instead of bt_gatt_attr* in deferred APIs

The initial deferred Read/Write response APIs used a raw
`struct bt_gatt_attr *` pointer. This is unsafe when services can be
dynamically registered or unregistered, and it also prevents future
support for safe local attribute access.

This patch replaces the pointer with an attribute handle, which matches
how remote ATT requests reference attributes and provides a stable
identifier for deferred responses.

Signed-off-by: Zhijie Zhong <zhongzhijie1@xiaomi.com>

Thalley commented on include/zephyr/bluetooth/gatt.h:
```diff
@@ -2035,6 +2065,32 @@ struct bt_gatt_read_params {
+ *  @retval -ENOTSUP EATT deferred responses not supported.
+ */
+int bt_gatt_send_read_rsp(struct bt_conn *conn, int err, uint16_t handle,
+			      const void *data, uint16_t length);
```
Indentation is off

Thalley commented on include/zephyr/bluetooth/gatt.h:
```diff
@@ -2035,6 +2065,32 @@ struct bt_gatt_read_params {
+ *  @retval -ENOMEM Out of memory.
+ *  @retval -ENOTSUP EATT deferred responses not supported.
+ */
+int bt_gatt_send_read_rsp(struct bt_conn *conn, int err, uint16_t handle,
```
I think @alwa-nordic has made a point that this API won't work for EATT.

jhedberg commented:
@zhongzhijie1 were you planning to continue this work and address the review comments? Seems like quite many people are waiting for the feature based on comments in #29270.
